### PR TITLE
Halo2 Early Support Implementation

### PR DESCRIPTION
**Pull request by ElusAegis**
### Halo2 configuration

We currently support Halo2 circuits in an experimental stage. To use Halo2, you need to set the `kind` to `halo2` in the `mopro-config.toml` file.

```toml
[circuit]
kind = "halo2" # Options: circom, halo2
dir = "mopro-core/examples/halo2/halo2-fibonacci" # Directory of the circuit
name = "fibonacci"                # Name of the circuit
```

The `dir` should point to the directory where the Halo2 circuit is located. The `name` should be the name of the circuit.

Note that currently the Halo2 circuit **must** be a cargo crate, with the package name hardcoded to `halo2-circuit`. This is due to us swapping out the default implementation of the `halo2-ciurcuit` crate with the user's circuit during the build process.
This is done using the `build.rs` script in the `mopro-core` crate, which changes the path to the default `examples/halo2/halo2-fibonacci` to `$dir`. This also requires the name of the package to be `halo2-circuit` for the substitution to work.

Additionally, the `halo2-circuit` crate **must** implement and expose three items, which you can see in the `mopro-core/examples/halo2/halo2-fibonacci/src/lib.rs`:
- `Circuit` - the struct that represents the circuit
- `prove` - the function that generates the proof
- `verify` - the function that verifies the proof

This is then used by the `mopro-core` crate to generate the proof and verify it.

### CODE REVIEW

File: HALO2.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+### Halo2 configuration
+
+We currently support Halo2 circuits in an experimental stage. To use Halo2, you need to set the `kind` to `halo2` in the `mopro-config.toml` file.
+
+```toml
+[circuit]
+kind = "halo2" # Options: circom, halo2
```
Comment (oskarth) [inline HALO2.md:7]: Could we rename this option to `adapter`? https://zkmopro.org/docs/intro

File: config-halo2.toml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# debug is for Rust library to be in debug mode and release for release mode
+# We recommend release mode by default for performance
+build_mode = "release" # Options: debug, release
+
+[circuit]
+kind = "halo2"                    # Proving system options: circom (default), halo2
+dir = "mopro-core/examples/halo2/halo2-fibonacci" # Directory of the circuit
+name = "fibonacci"                # Name of the circuit
+
+[dylib]
```
Comment (oskarth) [inline config-halo2.toml:18]: I guess these options are not needed for Halo2. Should we remove dylib and witness from this config?

File: mopro-core/Cargo.toml [hunk 3]
```diff
@@ -7,12 +7,25 @@ edition = "2021"
 
 [features]
 default = ["wasmer/dylib"]
+halo2 = ["halo2_proofs", "halo2-circuit"]
```
Comment (oskarth) [inline mopro-core/Cargo.toml:10]: Could we be consistent about either using `_` or `-`?

File: mopro-ffi/src/mopro.udl [hunk 4]
```diff
@@ -14,6 +14,13 @@ namespace mopro {
   [Throws=MoproError]
   boolean verify_proof2(bytes proof, bytes public_input);
 
+[Throws=MoproError]
```
Comment (oskarth) [inline mopro-ffi/src/mopro.udl:17]: (Indent)

Later on we could consider splitting this udl file up depending on flag, but I think this is fine for now!

**Review by oskarth — COMMENTED** [review]
Great start! Mostly minor comments.

Could we update options and docs here? Ideally it'd only init halo2 directory if that adapter is chosen (and likewise for Circom). Can also be done in follow up PR. https://github.com/zkmopro/mopro/tree/main/mopro-cli/src#mopro-cli

For merging this, I think as long as CI passes it is OK. Can also default to Halo2 turned off if there's more work to be done to make it work, to make it easier to integrate with other work.

File: templates/mopro-example-app/core/circuits/halo2-fibonacci/README.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# Halo2 Examples
+
+This repo includes a few simple examples to illustrate how to write circuit in Halo2.
+
+## Instruction
+
+Compile the repo
+
+```
+cargo build
```
Comment (vivianjeng) [inline templates/mopro-example-app/core/circuits/halo2-fibonacci/README.md:10]: this doesn't work in this directory
```sh
error[E0432]: unresolved import `halo2_examples`
 --> core/circuits/halo2-fibonacci/src/main.rs:9:5
  |
9 | use halo2_examples::{FinbonaciCircuit, write_keys, write_srs};
  |     ^^^^^^^^^^^^^^ use of undeclared crate or module `halo2_examples`

```

Comment (vivianjeng) [reply templates/mopro-example-app/core/circuits/halo2-fibonacci/README.md:10]: maybe outdated or need to be run in other directory?

**Review by vivianjeng — COMMENTED** [review]
Thank you for this PR
the example app works well for me!

File: HALO2.md (hunk 1 above)
Comment (ElusAegis) [reply HALO2.md:7]: Agree. Thank you for pointing out.

File: config-halo2.toml (hunk 2 above)
Comment (ElusAegis) [reply config-halo2.toml:18]: I tried to minimize changes to the original file structure. Will keep this for the improvement PR along with other more serious changes, like circom feature flags, etc.

File: mopro-core/Cargo.toml (hunk 3 above)
Comment (ElusAegis) [reply mopro-core/Cargo.toml:10]: I can switch `"halo2-circuit"` to `"halo2_circuit"` as this is a local dependency, as opposed to `"halo2_proofs"`, yet this seems to be a larger issue with rust dependencies using mostly `"-"` than `"_"`, hence this will stand out.  For example in the rest of the dependencies for circom, we also have both: `"once_cell"` and `"ark-ec"`. 

I propose to keep this until we rework the file to separate Circom dependencies from shared dependencies.

File: mopro-core/src/middleware/halo2/mod.rs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+const VK_BYTES: &[u8] = include_bytes!(env!("BUILD_VK_FILE"));
+
+static VK: Lazy<VerifyingKey<G1Affine>> = Lazy::new(|| {
+    let mut reader = Cursor::new(VK_BYTES);
+    VerifyingKey::read::<_, TargetCircuit<Fp>>(&mut reader, RawBytes).expect("Unable to read VK from file")
+});
+
+
+pub fn generate_halo2_proof2(
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply mopro-core/src/middleware/halo2/mod.rs:56]: There are two things here:

1. Regarding the suffix _"2"_ - I have used this to indicate similarity to a corresponding Circom method, however, I see that this will rather only cause confusion, hence I will change this.
2. Regarding the suffix _"halo2"_ instead of using `generate_proof` - I am not sure if the interface should be transparent to the proof system used, or rather fail explicitly when compiled for a wrong proof system by using a different function. I would love to hear your views on that, as well as @oskarth

File: mopro-core/src/middleware/mod.rs [hunk 7]
```diff
@@ -1,4 +1,8 @@
+#[cfg(not(feature = "halo2"))]
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply mopro-core/src/middleware/mod.rs:1]: Absolutely, right now I am keeping this off for future PR to simplify the content of this one.

File: mopro-ffi/src/circom.rs [hunk 8]
```diff
@@ -0,0 +1,473 @@
+#![allow(unused_variables)]
+
+use std::collections::HashMap;
+#[cfg(feature = "dylib")]
+use std::path::Path;
+
+#[cfg(not(feature = "halo2"))]
+use std::str::FromStr;
+#[cfg(not(feature = "halo2"))]
+use std::sync::RwLock;
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply mopro-ffi/src/circom.rs]: Nice, I have not seen this solution before, I will do this with the circom flag in the next few PRs.

File: scripts/cli/build_ios.sh [hunk 9]
```diff
@@ -74,6 +76,22 @@ determine_build_directory() {
     esac
 }
 
+# Select build features based on circuit type
+select_features() {
+    case $CIRCUIT_TYPE in
+        "circom")
+            CARGO_FEATURES=""
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply scripts/cli/build_ios.sh:83]: Right, I have missed this, thank you for the point. We will need to think how to manage this in the future

File: templates/mopro-example-app/core/circuits/halo2-fibonacci/README.md (hunk 5 above)
Comment (ElusAegis) [reply templates/mopro-example-app/core/circuits/halo2-fibonacci/README.md:10]: It was, fixed it, thank you for pointing out!

File: scripts/cli/build_ios.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (vivianjeng) [reply scripts/cli/build_ios.sh:83]: also here
can we create an issue here?

File: config-halo2.toml [hunk 10]
```diff
@@ -4,15 +4,22 @@
...
 [circuit]
+<<<<<<<< HEAD:config-halo2.toml
+adapter = "halo2"                    # Proving system options: circom (default), halo2
+dir = "mopro-core/examples/halo2/fibonacci" # Directory of the circuit
+name = "fibonacci"                # Name of the circuit
+========
+adapter = "circom"                              # Proving system options: circom (default), halo2
 dir = "mopro-core/examples/circom/keccak256" # Directory of the circuit
 name = "keccak256_256_test"                  # Name of the circuit
+>>>>>>>> mine/halo2-main:config-ci.toml
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply config-halo2.toml:23]: Sorry for my blunder. Fixed

File: mopro-core/examples/circom.rs [hunk 11]
```diff
@@ -1,7 +1,11 @@
+#[cfg(not(feature = "halo2"))]
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply mopro-core/examples/circom.rs:1]: Apparently you can not group `use` statements using braces outside of functions, so changed it to the following:

```rust
#[cfg(not(feature = "halo2"))]
use {mopro_core::middleware::circom::CircomState, num_bigint::BigInt, std::collections::HashMap};

```

Also did the same in `mopro-ffi`

File: mopro-core/examples/halo2/fibonacci/README.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+```
+cargo build
+```
+
+Run examples
+```
+cargo test -- --nocapture test_example1
+cargo test -- --nocapture test_example2
+cargo test -- --nocapture test_example3
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply mopro-core/examples/halo2/fibonacci/README.md]: Forgot about that readme, thank you, updated it.

**Comment by ElusAegis** [comment]
@vivianjeng thank you for the review and finding the issues. I have resolved them - ready to have another look

File: mopro-core/gpu_explorations/benchmarks/arkworks_pippenger_benchmark.txt [hunk 13]
```diff
@@ -1,9 +1,7 @@
 msm_size,num_msm,avg_processing_time(ms)
```
Comment (oskarth) [inline mopro-core/gpu_explorations/benchmarks/arkworks_pippenger_benchmark.txt:1]: Are these part of the PR? Benchmark stuff seems unrelated

**Review by oskarth — APPROVED** [review]
LGTM! Seems like some rebase is needed or so? I notice a few changes related to benchmarks and web that seem unrelated.

**Comment by ElusAegis** [comment]
@oskarth the changes to other files out of the scope come due to rebasing to main with MSM PR. Likely as I have merged using IDE rebase vs GitHub built-in rebase they are shown in the preview (https://github.com/orgs/community/discussions/51948), however, if you look into the commit history, they will be shown to have been made by other PRs.

File: mopro-core/gpu_explorations/benchmarks/arkworks_pippenger_benchmark.txt (hunk 13 above)
Comment (ElusAegis) [reply mopro-core/gpu_explorations/benchmarks/arkworks_pippenger_benchmark.txt:1]: Please see my reply here:


> @oskarth the changes to other files out of the scope come due to rebasing to main with MSM PR. Likely as I have merged using IDE rebase vs GitHub built-in rebase they are shown in the preview (https://github.com/orgs/community/discussions/51948), however, if you look into the commit history, they will be shown to have been made by other PRs.

**Comment by oskarth** [comment]
Right, it just requires a rebase to get a cleaner diff. If you just do it with normal git tools it shouldn't show up in any diff.

I already reviewed and approved PR. Let's wait for @vivianjeng if she had any other comments before merging.

**Review by vivianjeng — APPROVED** [review]
LGTM thank you

**Comment by vivianjeng** [comment]
@ElusAegis it still shows 
```sh
This branch cannot be rebased due to conflicts
```
maybe some branch got merged again

File: mopro-ffi/src/circom.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ElusAegis) [reply mopro-ffi/src/circom.rs]: Done in #172

File: mopro-core/src/middleware/halo2/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ElusAegis) [reply mopro-core/src/middleware/halo2/mod.rs]: In the end, I have renamed the proof and verify functions to be consistent with the circom naming: #173

File: mopro-ffi/src/circom.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ElusAegis) [reply mopro-ffi/src/circom.rs]: Sorry, did not create an issue, however, this PR (#173 ) addresses the problem with the FFI and attempts to improve the structure as well as does check that no mutually exclusive features exists.

File: scripts/cli/build_ios.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (ElusAegis) [reply scripts/cli/build_ios.sh:83]: Created it in #177

File: config-halo2.toml (hunk 2 above)
Comment (ElusAegis) [reply config-halo2.toml:18]: Created an issue for that, will try to do it some time next week
